src/Microsoft.Identity.Web/TokenAcquisition.cs

```diff
@@ -312,36 +338,56 @@ private IConfidentialClientApplication BuildConfidentialClientApplication()
                 request.PathBase,
                 microsoftIdentityOptions.CallbackPath.Value ?? string.Empty);
 
-            if (!applicationOptions.Instance.EndsWith("/"))
-                applicationOptions.Instance += "/";
-
-            string authority ;
-            IConfidentialClientApplication app = null;
-
-            if (microsoftIdentityOptions.IsB2C)
+           if (string.IsNullOrEmpty(applicationOptions.Instance) &&
+                !string.IsNullOrEmpty(microsoftIdentityOptions.Instance))
             {
-                authority = $"{applicationOptions.Instance}tfp/{microsoftIdentityOptions.Domain}/{microsoftIdentityOptions.DefaultUserFlow}";
-                app = ConfidentialClientApplicationBuilder
-                    .CreateWithApplicationOptions(applicationOptions)
-                    .WithRedirectUri(currentUri)
-                    .WithB2CAuthority(authority)
-                    .Build();
+                if (!microsoftIdentityOptions.Instance.EndsWith("/"))
+                    microsoftIdentityOptions.Instance += "/";
+                instance = microsoftIdentityOptions.Instance;
             }
             else
             {
-                authority = $"{applicationOptions.Instance}{applicationOptions.TenantId}/";
-                app = ConfidentialClientApplicationBuilder
-                    .CreateWithApplicationOptions(applicationOptions)
-                    .WithRedirectUri(currentUri)
-                    .WithAuthority(authority)
-                    .Build();
+                if (!applicationOptions.Instance.EndsWith("/"))
+                    applicationOptions.Instance += "/";
+                instance = applicationOptions.Instance;
             }
 
-            // Initialize token cache providers
-            _tokenCacheProvider?.InitializeAsync(app.AppTokenCache);
-            _tokenCacheProvider?.InitializeAsync(app.UserTokenCache);
+            string authority;
+            IConfidentialClientApplication app;
 
-            return app;
+            try
+            {
+                if (microsoftIdentityOptions.IsB2C)
```

Comment:
This piece of logic is not tested.

Reply:
we are missing a lot of tests. they will come in future PRs. 

---
In reply to: [385618752](https://github.com/AzureAD/microsoft-identity-web/pull/39#discussion_r385618752) [](ancestors = 385618752)